control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go

```diff
@@ -441,6 +441,16 @@ func (r *reconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result
 		errs = append(errs, r.client.Delete(ctx, nodeTuningCO))
 	}
 
+	// Delete the DNS operator deployment in the hosted cluster, if it is
+	// present there.  A separate DNS operator deployment runs as part of
+	// the hosted control-plane, but an upgraded cluster might still have
+	// an old DNS operator deployment in the hosted cluster.
+	dnsOperatorDeployment := manifests.DNSOperatorDeployment()
+	if err := r.client.Get(ctx, client.ObjectKeyFromObject(dnsOperatorDeployment), dnsOperatorDeployment); err == nil {
```

Comment:
Unfortunately it is not quite that easy, the `client` uses a cache to read against and in order to not cache the whole cluster, that cache has a labelselector which doesn't match for the dns-operator, which is why this code doesn't work and ultimately the upgrade test fails.

We have an uncached client (if you rebase, you will see it) but we don't want to use it all the time. The best I can think of is to use a `sync.Once` or some such to run this deletion that ends up entailing an api call once? If we do that, we can also just run the deletion unconditionally and tolerate an IsNotFound error

Reply:
Thanks!  I added a `sync.Once`.  I also needed to define a fake uncached client in `TestReconcileErrorHandling`.  I no longer need to modify `initialObjects`, so I deleted that change.